src/task4feedback/simulator/rl/models/env.py

```diff
@@ -0,0 +1,257 @@
+import copy
+import torch
+
+from collections import namedtuple
+from functools import partial
+from typing import Dict, List, Tuple
+from itertools import chain
+
+from ....logging import logger
+from ...utility import convert_to_float
+from ...task import SimulatedTask, SimulatedComputeTask
+from ....types import TaskState, TaskType, Device, Architecture
+from .globals import *
+
+
+def get_indegree(task: SimulatedTask) -> int:
+    return len(task.dependencies)
+
+
+def get_outdegree(task: SimulatedTask) -> int:
+    return len(task.dependents)
+
+
+class RLBaseEnvironment:
+    pass
+
+
+class RLEnvironment(RLBaseEnvironment):
+  """
+    *** RL state features ***
+
+    1. Global information (2):
+      * Completed tasks / total tasks
+      * Relative current wall-clock time (TODO(hc))
+
+    2. Task-specific information (4 + # types + # devices):
+      * In-degree / max in-degree
+      * Out-degree / max out-degree
+      * Task type ID (parameter)
+      * Expected exectuion time / max expected exectuion timm
+      * Depth (from a root to the current task)
+      * # parent tasks mapped per device / total # parent tasks
+
+    3. Device/network information (3 * # devices):
+      * Normalized resource usage (memory, VCUs) (TODO(hc))
+      * # mapped tasks per device / # active tasks
+      * Relative per-device idle time so far (TODO(hc))
+
+  """
+
+
+  def __init__(self, num_devices: int):
+      self.task_property_offset = 2
+      self.num_task_types = 4
+      self.task_property_len = 4 + self.num_task_types + num_devices
+      self.device_utilization_state_offset = self.task_property_offset + \
+                                             self.task_property_len
+      self.device_state_len = 3 * num_devices
+      self.state_dim = self.device_utilization_state_offset + \
+                       self.device_state_len
+      self.out_dim = num_devices
+      self.num_devices = num_devices
+      if logger.ENABLE_LOGGING:
+          logger.runtime.info(f"RL state dimension: {self.state_dim}.")
```

Comment:
Recommendation: Please use a new logger type (logger.reinforcement?) so its easy to debug/filter runtime issues (launching/mapping/reserving) separately from network training and testing.